pkg/operator/controller/machineconfig/crio.go

```diff
@@ -32,84 +33,77 @@ import (
+func (r *MachineConfigReconciler) enableCrioProf(ctx context.Context) (bool, error) {
+	createdNow := false
+	namespace := types.NamespacedName{Name: CrioProfilingConfigName}
 
-	// CrioServiceFile is the name of the CRI-O systemd service unit
-	CrioServiceFile = "crio.service"
+	err := r.createCrioProfConf(ctx)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		return createdNow, err
+	}
+	if err == nil {
+		createdNow = true
+	}
 
-	// CrioUnixSocketConfData contains the configuration required
-	// for enabling CRI-O profiling
-	CrioUnixSocketConfData = `[Service]
-Environment="ENABLE_PROFILE_UNIX_SOCKET=true"`
-)
+	// grace time for client cache to get refreshed
+	time.Sleep(500 * time.Millisecond)
 
-// ensureCrioProfConfigExists checks if CRI-O MachineConfig CR for
-// enabling profiling exists, if not creates the resource
-func (r *MachineConfigReconciler) ensureCrioProfConfigExists(ctx context.Context) (*mcv1.MachineConfig, bool, error) {
-	namespace := types.NamespacedName{Name: CrioProfilingConfigName}
-	criomc, exists, err := r.fetchCrioProfileConfig(ctx, namespace)
+	criomc, err := r.fetchCrioProfConf(ctx, namespace)
 	if err != nil {
-		return nil, false, err
+		return createdNow, fmt.Errorf("failed to ensure crio profiling config was indeed created: %w", err)
 	}
-	if !exists {
-		if err := r.createCrioProfileConfig(ctx); err != nil {
-			return nil, false, err
-		}
 
-		criomc, found, err := r.fetchCrioProfileConfig(ctx, namespace)
-		if err != nil || !found {
-			return nil, false, fmt.Errorf("failed to fetch just created crio config: %v", err)
-		}
-
-		return criomc, true, nil
+	r.Lock()
+	r.MachineConfig.PrevReconcileUpd["crio"] = MachineConfigInfo{
+		op:     "create",
+		config: criomc,
 	}
-	return criomc, false, nil
+	r.Unlock()
+
+	return createdNow, nil
 }
 
-// ensureCrioProfConfigNotExists checks if CRI-O MachineConfig CR for
+// disableCrioProf checks if CRI-O MachineConfig CR for
 // enabling profiling exists, if exists delete the resource
-func (r *MachineConfigReconciler) ensureCrioProfConfigNotExists(ctx context.Context) (bool, error) {
+func (r *MachineConfigReconciler) disableCrioProf(ctx context.Context) error {
 	namespace := types.NamespacedName{Name: CrioProfilingConfigName}
-	criomc, exists, err := r.fetchCrioProfileConfig(ctx, namespace)
+
+	criomc, err := r.fetchCrioProfConf(ctx, namespace)
+	if errors.IsNotFound(err) {
+		return nil
+	}
+
 	if err != nil {
-		return false, err
+		return err
 	}
-	if exists {
-		if err := r.deleteCrioProfileConfig(ctx, criomc); err != nil {
-			return false, err
-		}
 
-		_, exists, err = r.fetchCrioProfileConfig(ctx, namespace)
-		if err != nil || exists {
-			return false, fmt.Errorf("failed to ensure crio profiling config was indeed deleted: %v", err)
-		}
+	if err := r.deleteCrioProfConf(ctx, criomc); err != nil {
+		return err
+	}
 
-		return true, nil
+	r.Lock()
+	r.MachineConfig.PrevReconcileUpd["crio"] = MachineConfigInfo{
```

Comment:
This panic is possible here:
```
panic: assignment to entry in nil map

goroutine 1 [running]:

Program exited.
```
Please:
* Mitigate the potential of this panic
* use defer for r.Unlock() (defer is called even during a panic)

Reply:
Map is initialized [here](https://github.com/openshift/node-observability-operator/pull/45/files/aeb062be957c34efd81f81379613c81b00f2148a#diff-00a14626d7a56d8be3fe940ee9c4ffbdf0676de1f64453f29ae987c4f16961e3R149), when reconciler object is created and is not freed in the code, nil map panic should not be observed. But using defer is better approach, have incorporated comment.